### PR TITLE
[Merged by Bors] - feat(NumberTheory/NumberField): Restricting maps to the ring of integers

### PR DESCRIPTION
**Pull request by mbkybky**
This PR adds definitions of restricting maps to the ring of integers. That is, given a ring homomorphism (resp. ring isomorphism; resp. algebra homomorphism; resp. algebra isomorphism) `f` from`K` to `L`, we define the induced ring homomorphism (resp. ring isomorphism; resp. algebra homomorphism; resp. algebra isomorphism) from `𝓞 K` to `𝓞 L`, which is obtained by restricting `f` to `𝓞 K`.

Co-authored-by: Yongle Hu @mbkybky
Co-authored-by: Yuyang Zhao @FR-vdash-bot

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/Ring/Equiv.lean [hunk 1]
```diff
@@ -814,6 +814,33 @@ theorem symm_trans_self (e : R ≃+* S) : e.symm.trans e = RingEquiv.refl S :=
 
 end RingEquiv
 
+namespace RingEquiv
+
+variable [NonAssocSemiring R] [NonAssocSemiring S]
+
+/-- If a ring homomorphism has an inverse, it is a ring isomorphism. -/
+@[simps]
+def ofRingHom (f : R →+* S) (g : S →+* R) (h₁ : f.comp g = RingHom.id S)
```
Comment (riccardobrasca) [inline Mathlib/Algebra/Ring/Equiv.lean:823]: This is essentially [RingHom.inverse](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Ring/Equiv.html#RingHom.inverse), I am not sure we want to duplicate it. Do we do it for other types of homs?

**Review by riccardobrasca — COMMENTED** [review]
We already have [NumberField.inst_ringOfIntegersAlgebra](https://leanprover-community.github.io/mathlib4_docs/Mathlib/NumberTheory/NumberField/Basic.html#NumberField.inst_ringOfIntegersAlgebra). I see it's not exactly the same thing, but it is very close. What do you think?

File: Mathlib/Algebra/Ring/Equiv.lean (hunk 1 above)
Comment (mbkybky) [reply Mathlib/Algebra/Ring/Equiv.lean:823]: > This is essentially [RingHom.inverse](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Ring/Equiv.html#RingHom.inverse), I am not sure we want to duplicate it. Do we do it for other types of homs?

I used [AlgEquiv.ofAlgHom](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Algebra/Equiv.html#AlgEquiv.ofAlgHom) to construct `mapAlgEquiv`, and then I wanted to construct `mapRingEquiv` in a similar way. Therefore, I wrote `RingEquiv.ofRingHom` by imitating `AlgEquiv.ofAlgHom`.

**Comment by mbkybky** [comment]
> We already have [NumberField.inst_ringOfIntegersAlgebra](https://leanprover-community.github.io/mathlib4_docs/Mathlib/NumberTheory/NumberField/Basic.html#NumberField.inst_ringOfIntegersAlgebra). I see it's not exactly the same thing, but it is very close. What do you think?

Yes. What I really need is `mapAlgEquiv`. To this end, I wrote `mapRingHom` following the construction of `NumberField.inst_ringOfIntegersAlgebra`, and used it to write `mapAlgHom` and `mapAlgEquiv`.

**Comment by riccardobrasca** [comment]
Yes, the equiv version is definitely not there. What about avoiding `mapRingHom` (since it is just the `algebraMap` given by the algebra instance) but keeping the equiv?

**Comment by riccardobrasca** [comment]
Actually we want that for a tower of fields `M/L/K` (with `IsScalarTower ..`) the corresponing algebras of ring of integers satisfy `IsScalarTower`. Then proving your result should be easy (just make `L/K/L` using the ring hom and its inverse).

**Comment by mbkybky** [comment]
> Yes, the equiv version is definitely not there. What about avoiding `mapRingHom` (since it is just the `algebraMap` given by the algebra instance) but keeping the equiv?

I think `mapRingHom` would be more general. For example, if I have a ring homomorphism `L →+* L`, the induced `Algebra` instance will conflict with the instance induced by `RingHom.id`.

**Comment by riccardobrasca** [comment]
What I am saying is that if we have `L` a `K` algebra, there with your construction two ways of getting a ring hom `𝓞 K → 𝓞 L`: one is `algebraMap (𝓞 K) (𝓞 L)`, and the other your construction. It's not the end of the world (and you're right, your construction works given any map, without the `Algebra` instance), but it would be nice to have them defeq (so you should modify your construction to be `algebraMap`, after some `let _ : Algebra .. ..:=`). or at least we want the theorem saying they're the same.

**Comment by mbkybky** [comment]
> What I am saying is that if we have `L` a `K` algebra, there with your construction two ways of getting a ring hom `𝓞 K → 𝓞 L`: one is `algebraMap (𝓞 K) (𝓞 L)`, and the other your construction. It's not the end of the world (and you're right, your construction works given any map, without the `Algebra` instance), but it would be nice to have them defeq (so you should modify your construction to be `algebraMap`, after some `let _ : Algebra .. ..:=`). or at least we want the theorem saying they're the same.

I understand. But would it be a better to modify the definition of `NumberField.inst_ringOfIntegersAlgebra` (using `mapRingHom`)?  It's currently induced by a ring homomorphism, and inducing a ring homomorphism from it again might be too indirect.

**Comment by riccardobrasca** [comment]
> > What I am saying is that if we have `L` a `K` algebra, there with your construction two ways of getting a ring hom `𝓞 K → 𝓞 L`: one is `algebraMap (𝓞 K) (𝓞 L)`, and the other your construction. It's not the end of the world (and you're right, your construction works given any map, without the `Algebra` instance), but it would be nice to have them defeq (so you should modify your construction to be `algebraMap`, after some `let _ : Algebra .. ..:=`). or at least we want the theorem saying they're the same.
> 
> I understand. But would it be a better to modify the definition of `NumberField.inst_ringOfIntegersAlgebra` (using `mapRingHom`)? It's currently induced by a ring homomorphism, and inducing a ring homomorphism from it again might be too indirect.

Why not, but make sure to keep [this](https://github.com/leanprover-community/mathlib4/blob/367f25d3750097169f52c46b90f06497586b3a31/Mathlib/NumberTheory/NumberField/Basic.lean#L149) example working.

**Comment by mbkybky** [comment]
> > > What I am saying is that if we have `L` a `K` algebra, there with your construction two ways of getting a ring hom `𝓞 K → 𝓞 L`: one is `algebraMap (𝓞 K) (𝓞 L)`, and the other your construction. It's not the end of the world (and you're right, your construction works given any map, without the `Algebra` instance), but it would be nice to have them defeq (so you should modify your construction to be `algebraMap`, after some `let _ : Algebra .. ..:=`). or at least we want the theorem saying they're the same.
> > 
> > 
> > I understand. But would it be a better to modify the definition of `NumberField.inst_ringOfIntegersAlgebra` (using `mapRingHom`)? It's currently induced by a ring homomorphism, and inducing a ring homomorphism from it again might be too indirect.
> 
> Why not, but make sure to keep [this](https://github.com/leanprover-community/mathlib4/blob/367f25d3750097169f52c46b90f06497586b3a31/Mathlib/NumberTheory/NumberField/Basic.lean#L149) example working.

OK. I will try it tomorrow. Thank you!

**Comment by mbkybky** [comment]
!bench

**Review by riccardobrasca — APPROVED** [review]
Thanks!

bors merge

**Pull request closed without merging**